neo/Prompt/Commands/SC.py

```diff
@@ -283,7 +284,25 @@ def execute(self, arguments):
                 print("Incorrect password")
                 return False
 
-            return InvokeContract(wallet, tx, Fixed8.Zero(), from_addr=from_addr)
+            n_fee = Fixed8.Zero()
+            if priority_fee is not None:
+                n_fee = priority_fee
+                if n_fee is False:
+                    logger.debug("invalid fee")
+                    return False
+            print(f"Deploying with fee: {n_fee.ToString()}")
+
+            # calculate the required network fee for the tx and compare to n_fee
+            if tx.Size() > settings.MAX_FREE_TX_SIZE:
+                req_fee = Fixed8.FromDecimal(settings.FEE_PER_EXTRA_BYTE * (tx.Size() - settings.MAX_FREE_TX_SIZE))
+                if req_fee < settings.LOW_PRIORITY_THRESHOLD:
+                    req_fee = settings.LOW_PRIORITY_THRESHOLD
+                if n_fee < req_fee:
+                    print(f'Transaction cancelled. The tx size ({tx.Size()}) exceeds the max free tx size ({settings.MAX_FREE_TX_SIZE}).')
+                    print(f'A network fee of {req_fee.ToString()} GAS is required.')
+                    return False

```

Comment:
The overall checks here look good to me. We have to think about a more generic place to put this stuff though. Your suggestion of `wallet.MakeTransaction` makes sense as that appears (note: has to be validated) to be called before all type of transactions (Invoke, Claim etc). 

If we put it there we need to 
- excluded the check for `ClaimTransactions` according to https://github.com/neo-project/neo-plugins/blob/d35297618006bc0dfdae3bacab22e98df682d84b/SimplePolicy/SimplePolicyPlugin.cs#L101
It would probably be best to add this to the config file so it can be easily updated in the future when NEO decides to add more types of TX's to it.
- make sure we updated the docstrings
- have a new exception type for more control. e.g. `TransactionFeeException`. Otherwise we can't provide proper feedback to the user. Without it something like the following will just print insufficient funds, instead of telling the user that they need to increase their TX fee.
https://github.com/CityOfZion/neo-python/blob/e561e5b57d4093a9ee0c87a8312d4a4d16f6fd5e/neo/Prompt/Commands/Invoke.py#L52-L56

Reply:
Responses:
1) If you do a search for 'MakeTransaction' on GH, it looks like the only time `MakeTransaction` is used is associated with `Invoke` and `Contract` transactions. `Claim` transactions do not use `MakeTransaction` (see https://github.com/CityOfZion/neo-python/blob/e561e5b57d4093a9ee0c87a8312d4a4d16f6fd5e/neo/Prompt/Commands/Wallet.py#L278), so I think that issue is solved.
2) _blushing_ what are docstrings?
3) And I will try to see how create a new Exception type. I think I will model it after the special exceptions in JsonRpcApi.py

Reply:
1. sounds good then!
2. https://www.pythonforbeginners.com/basics/python-docstrings
3. it's pretty straightforward e.g.
```
class UnsufficientTXFee(Exception):
   pass
```
and then you can do i.e. `raise UnsufficientTXFee("require x, got y")`